### ISSUE TITLE
startOperation/endOperation only work with a single operation stack even across multiple CM instances

### ISSUE THREAD

**Issue by takamori** [codemirror/CodeMirror#5617]
With two independent CodeMirror instances, if one runs startOperation() on both CM instances, and finishes them in non-FIFO order, then a JS error will be thrown "Error: Cannot read property 'ownsGroup' of null"

Repro: https://stackblitz.com/edit/js-hjcwff?file=index.js
Click on the start button, then end operation on CM1 before CM2.

It does not repro if you end operation on CM2 before CM1.

I suspect that this is because operationGroup is a global instance here:
https://github.com/codemirror/CodeMirror/blob/master/src/util/operation_group.js#L3

**Comment by takamori** [comment]
I meant that in combination with how endOperation wipes the curOp value from all CM instances referenced by the group is what can cause this: https://github.com/codemirror/CodeMirror/blob/master/src/display/operations.js#L49

**Comment by marijnh** [comment]
The operation combining is intentional—it works that way to reduce browser relayouts when working with multiple editors.

Are you using `startOperation` and `endOperation` asynchronously? And if so, why? It'd make sense to make `endOperation` a no-op when there's no active operation, but depending on what you're doing, that might still not help.

**Comment by takamori** [comment]
Yes, it is asynchronous right now, as noted as a potential use case in the description of startOperation/endOperation ("if you want to buffer operations happening asynchronously..."). The intent in this case was to reduce the number of paints when doing a lot of rapid updates. Making endOperation a no-op would be harmless for our use case (at least, compared to having an error thrown... :) ).

But given the overall behavior, we're looking into whether we can avoid the asynchronous ordering and perform updates synchronously instead.

**Comment by marijnh** [comment]
Attached patch fixes the crash. Does that solve the issue for you?

**Comment by takamori** [comment]
Sorry for the late response. I believe it would have, but it looks like we've already made the change I mentioned earlier as being investigated.
